test/py/test_migration.py

```diff
@@ -28,139 +29,504 @@
+
+    transition_to_state(VFIO_USER_DEVICE_STATE_STOP_COPY)
+    read_data = data
+    result = msg(ctx, sock, VFIO_USER_MIG_DATA_READ, payload)
+    assert len(result) == argsz
+    assert result[len(vfio_user_mig_data()):] == data
+
+
+def test_handle_mig_data_read_too_long():
+    global read_data
+
+    transition_to_state(VFIO_USER_DEVICE_STATE_RUNNING)
+
+    payload = vfio_user_mig_data(
+        argsz=len(vfio_user_mig_data()) + 8,
+        size=8
+    )
+
+    # When we set up the tests at the top of this file we specify that the max
+    # data transfer size is 4 bytes. Here we test to check that a transfer of 8
+    # bytes fails.
+
+    transition_to_state(VFIO_USER_DEVICE_STATE_PRE_COPY)
+    read_data = bytes([1, 2, 3, 4, 5, 6, 7, 8])
+    msg(ctx, sock, VFIO_USER_MIG_DATA_READ, payload, expect=errno.EINVAL)
+
+
+def test_handle_mig_data_read_invalid_state():
+    global read_data
+
+    payload = vfio_user_mig_data(
+        argsz=len(vfio_user_mig_data()) + 4,
+        size=4
+    )
+
+    data = bytes([1, 2, 3, 4])
+
+    transition_to_state(VFIO_USER_DEVICE_STATE_RUNNING)
+    read_data = data
+    msg(ctx, sock, VFIO_USER_MIG_DATA_READ, payload, expect=errno.EINVAL)
+
+    transition_to_state(VFIO_USER_DEVICE_STATE_STOP)
+    read_data = data
+    msg(ctx, sock, VFIO_USER_MIG_DATA_READ, payload, expect=errno.EINVAL)
 
-    ret = vfu_device_quiesced(ctx, 0)
-    assert ret == 0
 
-    print("waiting for reply")
-    get_reply(sock, errno.ENOTTY)
-    print("received reply")
+def test_handle_mig_data_read_failed_callback():
+    global fail_callbacks
+
+    transition_to_state(VFIO_USER_DEVICE_STATE_PRE_COPY)
+
+    fail_callbacks = True
+
+    payload = vfio_user_mig_data(
+        argsz=len(vfio_user_mig_data()) + 4,
+        size=4
+    )
+
+    msg(ctx, sock, VFIO_USER_MIG_DATA_READ, payload, expect=errno.EINVAL)
+
+    fail_callbacks = False
+
+
+def test_handle_mig_data_read_short_write():
+    payload = struct.pack("I", 8)
+
+    msg(ctx, sock, VFIO_USER_MIG_DATA_READ, payload, expect=errno.EINVAL)
+
+
+def test_handle_mig_data_write():
+    payload = vfio_user_mig_data(
+        argsz=len(vfio_user_mig_data()) + 4,
+        size=4
+    )
+
+    data = bytes([1, 2, 3, 4])
+
+    transition_to_state(VFIO_USER_DEVICE_STATE_RESUMING)
+    msg(ctx, sock, VFIO_USER_MIG_DATA_WRITE, bytes(payload) + data)
+    assert write_data == data
+
+
+def test_handle_mig_data_write_too_long():
+    payload = vfio_user_mig_data(
+        argsz=len(vfio_user_mig_data()) + 8,
+        size=8
+    )
+
+    # When we set up the tests at the top of this file we specify that the max
+    # data transfer size is 4 bytes. Here we test to check that a transfer of 8
+    # bytes fails.
+
+    data = bytes([1, 2, 3, 4, 5, 6, 7, 8])
+    transition_to_state(VFIO_USER_DEVICE_STATE_RESUMING)
+    msg(ctx, sock, VFIO_USER_MIG_DATA_WRITE, bytes(payload) + data,
+        expect=errno.EINVAL)
+
+
+def test_handle_mig_data_write_invalid_state():
```

Comment:
nit: move before `test_handle_mig_data_write_too_long` so that all tests that test state are grouped together